### PR TITLE
[New protocol] Add WMI support

### PR DESCRIPTION
**Pull request by XiaoliChan**
### Reason why: 
In some situations, the firewall may disable port 445 traffic and only allow port 135 traffic incoming

With SMB enumeration
image

Without SMB enumeration
image

### Features
query function(from smb.py)
image

execute function(get command result from the registry, don't need SMB #59)
image

### Resources
[Orange-Cyberdefense's cme-wmi](https://github.com/Orange-Cyberdefense/cme-wmi)
[Impacket's wmiexec.py](https://github.com/SecureAuthCorp/impacket/blob/master/examples/wmiexec.py)
[Impacket's wmiquery.py](https://github.com/SecureAuthCorp/impacket/blob/master/examples/wmiquery.py)
[wmiexec-RegOut](https://github.com/XiaoliChan/wmiexec-RegOut)

### CODE REVIEW

**Comment by XiaoliChan** [comment]
@mpgn All things have been done now, I hope you can test it 😋

**Comment by mpgn** [comment]
Yeah I was waiting for you to tell me :)

**Comment by XiaoliChan** [comment]
> Yeah I was waiting for you to tell me :)

Cool, I think I need to open a PR again because now it looks very messy.

File: cme/protocols/wmi.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            dcom = DCOMConnection(self.host, self.username, self.password, self.domain, self.lmhash, self.nthash, oxidResolver=False)
+            iInterface = dcom.CoCreateInstanceEx(CLSID_WbemLevel1Login, IID_IWbemLevel1Login)
+        except Exception as e:
+            if "rpc_s_access_denied" in str(e):
+                return True
+        return False
+    
+    def enum_host_info(self):
+        # smb no open, specify the domain
+        if self.args.domain:
```
Comment (mpgn) [inline cme/protocols/wmi.py:81]: add a flag `--no-smb`

File: cme/protocols/wmi.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                logging.error("Error retrieving host domain: {} specify one manually with the '-d' flag".format(e))
+
+            if self.args.domain:
+                self.domain = self.args.domain
+
+            if self.args.local_auth:
+                self.domain = self.hostname
+
+    def print_host_info(self):
+        if self.args.domain:
```
Comment (mpgn) [inline cme/protocols/wmi.py:114]: same

File: cme/protocols/wmi/wmiexec_regout.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    def execute_remote(self, data):
+        # Save result as txt file
+        resultTXT = "C:\\windows\\temp\\" + str(uuid.uuid4()) + ".txt"
+        self.logger.highlight("[+] Executing command: \" %s \""%data)
+        data = data + " > " + resultTXT
+        command = self.__shell + self.encodeCommand(data)
+        self.__win32Process.Create(command, self.__pwd, None)
+        self.logger.highlight("[+] Waiting 5 for command completely executed.")
+        time.sleep(5)
```
Comment (mpgn) [inline cme/protocols/wmi/wmiexec_regout.py:54]: timeout should be an option that can be configured by the user

**Comment by mpgn** [comment]
When I try with -x or -q option, cme never finish can you look at it ?

image

A stacktrace: 

image

File: cme/protocols/wmi.py (hunk 1 above)
Comment (XiaoliChan) [reply cme/protocols/wmi.py:81]: That is what I need👍

File: cme/protocols/wmi/wmiexec_regout.py (hunk 3 above)
Comment (XiaoliChan) [reply cme/protocols/wmi/wmiexec_regout.py:54]: Sure

**Comment by XiaoliChan** [comment]
> When I try with -x or -q option, cme never finish can you look at it ?
> 
> image
> 
> Have no idea why it will stuck there, probably something wrong in "remove temporary registry key"

**Comment by XiaoliChan** [comment]
@mpgn I will try to fix the bugs next week 👀

**Comment by XiaoliChan** [comment]
@mpgn everything looks good now, I hope you can test it again :)

**Comment by mpgn** [comment]
Can you make the PR on https://github.com/mpgn/CrackMapExec ? This repo is not supported anymore

**Comment by XiaoliChan** [comment]
> Can you make the PR on https://github.com/mpgn/CrackMapExec ? This repo is not supported anymore

Sure

**Pull request closed without merging**